pkg/reconciler/pipelinerun/pipelinerun_expired.go

```diff
@@ -0,0 +1,145 @@
+	if expired, err := tc.processPrTTL(pr); err != nil {
+		return err
+	} else if !expired {
+		return nil
+	}
+
+	// The PipelineRun's TTL is assumed to have expired, but the PipelineRun TTL might be stale.
+	// Before deleting the PipelineRun, do a final sanity check.
+	// If TTL is modified before we do this check, we cannot be sure if the TTL truly expires.
+	// The latest PipelineRun may have a different UID, but it's fine because the checks will be run again.
+	fresh, err := tc.PipelineClientSet.TektonV1alpha1().PipelineRuns(namespace).Get(name, metav1.GetOptions{})
+	if errors.IsNotFound(err) {
+		return nil
+	}
+	if err != nil {
+		return err
+	}
+	// Use the latest PipelineRun TTL to see if the TTL truly expires.
+	if expired, err := tc.processPrTTL(fresh); err != nil {
+		return err
+	} else if !expired {
+		return nil
+	}
+	// Cascade deletes the PipelineRuns if TTL truly expires.
+	policy := metav1.DeletePropagationForeground
+	options := &metav1.DeleteOptions{
+		PropagationPolicy: &policy,
+		Preconditions:     &metav1.Preconditions{UID: &fresh.UID},
+	}
+	tc.Logger.Infof("Cleaning up PipelineRun %s/%s", namespace, name)
+
+	return tc.PipelineClientSet.TektonV1alpha1().PipelineRuns(fresh.Namespace).Delete(fresh.Name, options)
+}
+
+// processTTL checks whether a given PipelineRun's TTL has expired, and add it to the queue after the TTL is expected to expire
+// if the TTL will expire later.
+func (tc *Reconciler) processPrTTL(pr *apispipeline.PipelineRun) (expired bool, err error) {
+	// We don't care about the PipelineRuns that are going to be deleted, or the ones that don't need clean up.
+	if pr.DeletionTimestamp != nil || !pipelineRunCleanup(pr) {
+		return false, nil
+	}
+
+	now := tc.clock.Now()
+	t, err := tc.prTimeLeft(pr, &now)
+	if err != nil {
+		return false, err
+	}
+
+	// TTL has expired
+	if *t <= 0 {
+		return true, nil
+	}
+
+	controller.NewImpl(tc, tc.Logger, ControllerName).EnqueueAfter(pr, *t)
+	return false, nil
+}
+
+func getPrFinishAndExpireTime(pr *apispipeline.PipelineRun) (*time.Time, *time.Time, error) {
+	if !pipelineRunCleanup(pr) {
+		return nil, nil, fmt.Errorf("PipelineRun %s/%s should not be cleaned up", pr.Namespace, pr.Name)
+	}
+	finishAt, err := pipelineRunFinishTime(pr)
+	if err != nil {
+		return nil, nil, err
+	}
+	finishAtUTC := finishAt.Inner.UTC()
+	expireAtUTC := finishAtUTC.Add(pr.Spec.ExpirationSecondsTTL.Duration)
+	return &finishAtUTC, &expireAtUTC, nil
+}
+
+func (tc *Reconciler) prTimeLeft(pr *apispipeline.PipelineRun, since *time.Time) (*time.Duration, error) {
+	finishAt, expireAt, err := getPrFinishAndExpireTime(pr)
+	if err != nil {
+		return nil, err
+	}
+	if finishAt.UTC().After(since.UTC()) {
+		tc.Logger.Warnf("Warning: Found PipelineRun %s/%s succeeded in the future. This is likely due to time skew in the cluster. PipelineRun cleanup will be deferred.", pr.Namespace, pr.Name)
```

Comment:
I suggest adding a comment here that points to `https://kubernetes.io/docs/concepts/workloads/controllers/ttlafterfinished/#time-skew`